From 392880527193efc735116c2132216a08884c5229 Mon Sep 17 00:00:00 2001
From: Jun Zhang <jun.zhang@arm.com>
Date: Fri, 30 Jun 2023 10:50:11 +0800
Subject: [PATCH] feat(rdkronos): Add support for OP-TEE SPMC

These changes are to add support for loading and booting
OP-TEE as SPMC running at SEL1 for rdkronos platform.

Upstream-Status: Pending
Signed-off-by: Jun Zhang <jun.zhang@arm.com>
Signed-off-by: Jiamei Xie <jiamei.xie@arm.com>

---
 .../rdkronos/fdts/rdkronos_fw_config.dts      |  6 ++++
 .../fdts/rdkronos_optee_spmc_manifest.dts     | 30 +++++++++++++++++++
 .../arm/board/rdkronos/include/platform_def.h |  9 +++++-
 plat/arm/board/rdkronos/platform.mk           | 10 ++++++-
 plat/arm/board/rdkronos/rdkronos_plat.c       |  1 +
 5 files changed, 54 insertions(+), 2 deletions(-)
 create mode 100644 plat/arm/board/rdkronos/fdts/rdkronos_optee_spmc_manifest.dts

diff --git a/plat/arm/board/rdkronos/fdts/rdkronos_fw_config.dts b/plat/arm/board/rdkronos/fdts/rdkronos_fw_config.dts
index 7190c484e..00c435ef6 100644
--- a/plat/arm/board/rdkronos/fdts/rdkronos_fw_config.dts
+++ b/plat/arm/board/rdkronos/fdts/rdkronos_fw_config.dts
@@ -18,6 +18,12 @@
 			id = <TB_FW_CONFIG_ID>;
 		};
 
+		tos_fw-config {
+			load-address = <0x0 0x70000>;
+			max-size = <0x1000>;
+			id = <TOS_FW_CONFIG_ID>;
+		};
+
 		hw-config {
 			load-address = <0x0 0x83000000>;
 			max-size = <0x8000>;
diff --git a/plat/arm/board/rdkronos/fdts/rdkronos_optee_spmc_manifest.dts b/plat/arm/board/rdkronos/fdts/rdkronos_optee_spmc_manifest.dts
new file mode 100644
index 000000000..0b7f18d1d
--- /dev/null
+++ b/plat/arm/board/rdkronos/fdts/rdkronos_optee_spmc_manifest.dts
@@ -0,0 +1,30 @@
+/*
+ * Copyright (c) 2023 Arm Limited. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+/dts-v1/;
+
+/ {
+	compatible = "arm,ffa-core-manifest-1.0";
+	#address-cells = <2>;
+	#size-cells = <1>;
+
+	/*
+	 * BL32 image details needed by SPMC
+	 *
+	 * Note:
+	 * binary_size: size of BL32 + TOS_FW_CONFIG
+	 */
+
+	attribute {
+		spmc_id = <0x8000>;
+		maj_ver = <0x1>;
+		min_ver = <0x0>;
+		exec_state = <0x0>;
+		load_address = <0x0 0xfe000000>;
+		entrypoint = <0x0 0xfe000000>;
+		binary_size = <0x02000000>;
+	};
+};
+
diff --git a/plat/arm/board/rdkronos/include/platform_def.h b/plat/arm/board/rdkronos/include/platform_def.h
index bf3163a93..c9bc27e2a 100644
--- a/plat/arm/board/rdkronos/include/platform_def.h
+++ b/plat/arm/board/rdkronos/include/platform_def.h
@@ -35,6 +35,9 @@
 #define PLAT_CSS_MHU_BASE		UL(0x2A920000)
 #define PLAT_ARM_NSTIMER_FRAME_ID	U(0)
 
+#define PLAT_ARM_SPMC_BASE		UL(0xfe000000)
+#define PLAT_ARM_SPMC_SIZE		UL(0x02000000)
+
 #define SOC_CSS_SEC_UART_BASE		UL(0x2A410000)
 #define SOC_CSS_NSEC_UART_BASE		UL(0x2A400000)
 #define SOC_CSS_UART_SIZE		UL(0x10000)
@@ -101,9 +104,13 @@
 						MT_DEVICE | MT_RO | MT_SECURE)
 
 #define KRONOS_MAP_NS_DRAM1	MAP_REGION_FLAT(ARM_DRAM1_BASE,	\
-						ARM_DRAM1_SIZE,	\
+						ARM_DRAM1_SIZE - PLAT_ARM_SPMC_SIZE,	\
 						MT_MEMORY | MT_RW | MT_NS)
 
+#define KRONOS_MAP_S_DRAM1	MAP_REGION_FLAT(PLAT_ARM_SPMC_BASE, \
+						PLAT_ARM_SPMC_SIZE,	\
+						MT_MEMORY | MT_RW | MT_SECURE)
+
 #define KRONOS_DEVICE_BASE	(0x20000000)
 #define KRONOS_DEVICE_SIZE	(0x20000000)
 #define KRONOS_MAP_DEVICE	MAP_REGION_FLAT(KRONOS_DEVICE_BASE, \
diff --git a/plat/arm/board/rdkronos/platform.mk b/plat/arm/board/rdkronos/platform.mk
index 11d1e3906..e3bb2075d 100644
--- a/plat/arm/board/rdkronos/platform.mk
+++ b/plat/arm/board/rdkronos/platform.mk
@@ -21,7 +21,7 @@ CSS_USE_SCMI_SDS_DRIVER			:= 1
 ENABLE_FEAT_FGT				:= 1
 ENABLE_FEAT_ECV				:= 1
 HW_ASSISTED_COHERENCY			:= 1
-NEED_BL32				:= 0
+NEED_BL32				:= yes
 USE_COHERENT_MEM			:= 0
 
 GIC_ENABLE_V4_EXTN	:= 1
@@ -61,6 +61,14 @@ FW_CONFIG		:=	${BUILD_PLAT}/fdts/${PLAT}_fw_config.dtb
 TB_FW_CONFIG		:=	${BUILD_PLAT}/fdts/${PLAT}_tb_fw_config.dtb
 HW_CONFIG		:=	${BUILD_PLAT}/fdts/${PLAT}.dtb
 
+RDKRONOS_SPMC_MANIFEST_DTS	:=	${RDKRONOS_BASE}/fdts/${PLAT}_optee_spmc_manifest.dts
+FDT_SOURCES		+=	${RDKRONOS_SPMC_MANIFEST_DTS}
+RDKRONOS_TOS_FW_CONFIG	:=	${BUILD_PLAT}/fdts/${PLAT}_optee_spmc_manifest.dtb
+
+# Add the TOS_FW_CONFIG to FIP and specify the same to certtool
+$(eval $(call TOOL_ADD_PAYLOAD,${RDKRONOS_TOS_FW_CONFIG},--tos-fw-config,${RDKRONOS_TOS_FW_CONFIG}))
+
+
 # Add the FW_CONFIG to FIP and specify the same to certtool
 $(eval $(call TOOL_ADD_PAYLOAD,${FW_CONFIG},--fw-config,${FW_CONFIG}))
 # Add the TB_FW_CONFIG to FIP and specify the same to certtool
diff --git a/plat/arm/board/rdkronos/rdkronos_plat.c b/plat/arm/board/rdkronos/rdkronos_plat.c
index 1b8dd8a37..db3ab9e84 100644
--- a/plat/arm/board/rdkronos/rdkronos_plat.c
+++ b/plat/arm/board/rdkronos/rdkronos_plat.c
@@ -20,6 +20,7 @@ const mmap_region_t plat_arm_mmap[] = {
 #if IMAGE_BL2
 	KRONOS_MAP_NS_DRAM1,
 #endif
+	KRONOS_MAP_S_DRAM1,
 	{0}
 };
 
-- 
2.25.1

